packages/overlay/src/integrations/sentry/hooks/useTraceData.tsx

```diff
@@ -0,0 +1,69 @@
+import { useCallback, useMemo, useState } from "react";
+import { useLocation, useNavigate } from "react-router-dom";
+import { useSpotlightContext } from "~/lib/useSpotlightContext";
+import { useSentryTraces } from "../data/useSentrySpans";
+import useTraceFiltering from "./useTraceFiltering";
+
+/**
+ * hook that manages all trace-related data and logic
+ * including filtering, searching, and visibility control via URL parameters
+ */
+export function useTraceData() {
+  const context = useSpotlightContext();
+  const location = useLocation();
+  const navigate = useNavigate();
+
+  const { allTraces, localTraces } = useSentryTraces();
+
+  const searchParams = useMemo(() => new URLSearchParams(location.search), [location.search]);
+  const showNonLocalTraces =
+    searchParams.has("showNonLocalTraces") || !context.experiments["sentry:focus-local-events"];
+
+  const [searchQuery, setSearchQuery] = useState("");
+  const [activeFilters, setActiveFilters] = useState<string[]>([]);
+
+  const visibleTraces = useMemo(
+    () => (showNonLocalTraces ? allTraces : localTraces),
+    [showNonLocalTraces, allTraces, localTraces],
+  );
+
+  const { TRACE_FILTER_CONFIGS, filteredTraces } = useTraceFiltering(visibleTraces, activeFilters, searchQuery);
+
+  const nonLocalTraceCount = allTraces.length - localTraces.length;
+
+  const setShowNonLocalTraces = useCallback(
+    (show: boolean) => {
+      const newParams = new URLSearchParams(location.search);
+      if (show) {
+        newParams.set("showNonLocalTraces", "true");
+      } else {
+        newParams.delete("showNonLocalTraces");
+      }
+
+      const currentPath = location.pathname;
+      const newSearch = newParams.toString();
+      navigate(currentPath + (newSearch ? `?${newSearch}` : ""), { replace: true });
```

Comment:
If `location` is a `URL` object (and it looks like one), you should be able to do this instead:
```suggestion
      if (show) {
        location.searchParams.set("showNonLocalTraces", "true");
      } else {
        location.searchParams.delete("showNonLocalTraces");
      }

      navigate(location), { replace: true });
```

If it is not, then you can still do:
```ts
const newUrl = new URL(location);
newUrl.search = newParams.toString();
navigate(newUrl, { replace: true });
```